PushSharp.Core/PushServiceBase.cs

```diff
@@ -367,284 +362,284 @@ public TimeSpan AverageSendTime
-				}
-				else if (destroyed.HasValue && destroyed.Value)
-				{
-					var evt = this.OnChannelDestroyed;
-					if (evt != null)
-						evt(this);
-				}
-			}
-		}
-
-
-	    private long totalSendCount = 0;
-
-		private void DoChannelWork(IPushChannel channel, CancellationTokenSource cancelTokenSource)
-		{
+                            // newChannel.Dispose();
+                        }, TaskContinuationOptions.OnlyOnFaulted);
+
+                        channels.Add(chanWorker);
+
+                        newCount = channels.Count;
+                        destroyed = false;
+                    }
+                    else if (action == ChannelScaleAction.Destroy && channels.Count > 0)
+                    {
+                        var channelOn = channels[0];
+                        channels.RemoveAt(0);
+
+                        //Stop the channel worker, which will dispose the channel too
+                        channelOn.Dispose();
+
+                        newCount = channels.Count;
+                        destroyed = true;
+                    }
+                }
+
+                if (destroyed.HasValue && !destroyed.Value)
+                {
+                    var evt = OnChannelCreated;
+                    if (evt != null)
+                        evt(this, newChannel);
+                }
+                else if (destroyed.HasValue && destroyed.Value)
+                {
+                    var evt = this.OnChannelDestroyed;
+                    if (evt != null)
+                        evt(this);
+                }
+            }
+        }
+
+
+        private long totalSendCount = 0;
+
+        private void DoChannelWork(IPushChannel channel, CancellationTokenSource cancelTokenSource)
+        {
             long sendCount = 0;
 
-			while (!cancelTokenSource.IsCancellationRequested)
-			{
-				var notification = queuedNotifications.Dequeue (true);
-			    
-				if (notification == null)
-				{
-					Thread.Sleep(100);
-					continue;
-				}
-          //      Trace.WriteLine("Channel " + id + " processing notification " + notification.GetHashCode());
+            while (!cancelTokenSource.IsCancellationRequested)
+            {
+                var notification = queuedNotifications.Dequeue(true);
+
+                if (notification == null)
+                {
+                    Thread.Sleep(100);
+                    continue;
+                }
+                //      Trace.WriteLine("Channel " + id + " processing notification " + notification.GetHashCode());
                 ManualResetEvent waitForNotification = null;
 
                 if (this.BlockOnMessageResult)
                 {
                     waitForNotification = new ManualResetEvent(false);
                 }
 
-				var msWaited = (DateTime.UtcNow - notification.EnqueuedTimestamp).TotalMilliseconds;
+                var msWaited = (DateTime.UtcNow - notification.EnqueuedTimestamp).TotalMilliseconds;
+
+                lock (measurementsLock)
```

Comment:
i think we can remove this lock and the measurement, looks like it's not even used

Reply:
hm looks like is being used to calculate avg time and we do report that on that stats page